python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,623 @@
+    if (threadIdx.x + (item * blockDim.x) < size) {
+      T load = data[threadIdx.x + item * blockDim.x];
+      local_sum += load;
+    }
+  }
+
+  atomicAdds(sum, local_sum);
+
+  __syncthreads();
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ void device_var(
+  T const* data, int const items_per_thread, size_type size, T* sum, double* var)
+{
+  // Calculate how many elements each thread is working on
+  T local_sum      = 0;
+  double local_var = 0;
+  double mean;
+
+  device_sum<T>(data, items_per_thread, size, sum);
+
+  __syncthreads();
+
+  mean = (*sum) / static_cast<double>(size);
+
+// Calculate local sum for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (threadIdx.x + (item * blockDim.x) < size) {
+      T load      = data[threadIdx.x + item * blockDim.x];
+      double temp = load - mean;
+      temp        = pow(temp, 2);
+      local_var += temp;
+    }
+  }
+
+  atomicAdds(var, local_var);
+
+  __syncthreads();
+
+  *var = *var / (size - 1);
+
+  __syncthreads();
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ void device_max(
+  T const* data, int const items_per_thread, size_type size, T init_val, T* smax)
+{
+  T local_max = init_val;
+
+// Calculate local max for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (threadIdx.x + (item * blockDim.x) < size) {
+      T load    = data[threadIdx.x + item * blockDim.x];
+      local_max = max(local_max, load);
+    }
+  }
+
+  __syncthreads();
+
+  // Calculate local max for each group
+  atomicMax(smax, local_max);
+
+  __syncthreads();
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ void device_min(
+  T const* data, int const items_per_thread, size_type size, T init_val, T* smin)
+{
+  T local_min = init_val;
+
+// Calculate local min for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (threadIdx.x + (item * blockDim.x) < size) {
+      T load    = data[threadIdx.x + item * blockDim.x];
+      local_min = min(local_min, load);
+    }
+  }
+
+  __syncthreads();
+
+  // Calculate local min for each group
+  atomicMin(smin, local_min);
+
+  __syncthreads();
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ void device_idxmax(T const* data,
+                              int const items_per_thread,
+                              int64_t const* index,
+                              size_type size,
+                              T init_val,
+                              T* smax,
+                              int64_t* sidx)
+{
+  // Calculate how many elements each thread is working on
```

Comment:
This comment is wrong, since the calculation of items per thread is now done outside.